pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/functioncache/FunctionCacheManagerImpl.java

```diff
@@ -72,87 +70,70 @@ public void registerFunctionInstance(String fid,
+                    fid,
+                    new FunctionCacheEntry(
                             requiredJarFiles,
                             requiredClasspaths,
                             urls,
                             eid, rootClassLoader));
-                } catch (Throwable cause) {
-                    Exceptions.rethrowIOException(cause);
-                }
-            } else {
+            if (entry != null) {
                 entry.register(
-                    eid,
-                    requiredJarFiles,
-                    requiredClasspaths);
+                        eid,
+                        requiredJarFiles,
+                        requiredClasspaths);
             }
+        } catch (Throwable cause) {
+            Exceptions.rethrowIOException(cause);
         }
     }
 
     @Override
     public void registerFunctionInstanceWithArchive(String fid, String eid,
-                                                    String narArchive, String narExtractionDirectory) throws IOException {
+                                                    String narArchive,
+                                                    String narExtractionDirectory) throws IOException {
         if (fid == null) {
             throw new NullPointerException("FunctionID not set");
         }
-
-        synchronized (cacheFunctions) {
-            FunctionCacheEntry entry = cacheFunctions.get(fid);
-
+        // Create new cache entry.
+        try {
+            FunctionCacheEntry entry = cacheFunctions.putIfAbsent(fid, new FunctionCacheEntry(narArchive, eid,
+                    rootClassLoader, narExtractionDirectory));
             if (null != entry) {
                 entry.register(eid, Collections.singleton(narArchive), Collections.emptyList());
-                return;
-            }
-
-            // Create new cache entry
-            try {
-                cacheFunctions.put(fid, new FunctionCacheEntry(narArchive, eid, rootClassLoader, narExtractionDirectory));
-            } catch (Throwable cause) {
-                Exceptions.rethrowIOException(cause);
             }
+        } catch (Throwable cause) {
+            Exceptions.rethrowIOException(cause);
         }
     }
 
     @Override
-    public void unregisterFunctionInstance(String fid,
-                                           String eid) {
-        synchronized (cacheFunctions) {
-            FunctionCacheEntry entry = cacheFunctions.get(fid);
-
-            if (null != entry) {
-                if (entry.unregister(eid)) {
-                    cacheFunctions.remove(fid);
-                    entry.close();
-                }
+    public void unregisterFunctionInstance(String fid, String eid) {
+        FunctionCacheEntry entry = cacheFunctions.get(fid);
+        if (null != entry) {
+            if (entry.unregister(eid)) {
+                cacheFunctions.remove(fid);
+                entry.close();
             }
         }
```

Comment:
`get` and `remove` are also two atomic operations. We can use `remove` instead of them.

```suggestion
        FunctionCacheEntry entry = cacheFunctions.remove(fid);
        if (entry != null && entry.unregister(eid)) {
            entry.close();
        }
```

Reply:
Sorry my suggestion change here is wrong. It looks like we cannot `remove` it if `unregister` returns false.